feat(eslint-plugin): [prefer-nullish-coalescing] add ignoreTernaryTests option

jguddas opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

-   [x] Addresses an existing open issue: fixes #4905
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

I added an option to `prefer-nullish-coalescing` (`ignoreTernaryTests`) which when set to `false` highlights ternary expressions that can be converted to use the nullish coalescing operator (`??`) instead.

If I have forgotten anything, or you have suggestions what I can do better please let me know, this is my first contribution here 😄

tduyduc commented:
How about the `==` operator where `null` and `undefined` are equated (for example: `null != x ? x : 'a default value'`)?

tduyduc commented on packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md:
```diff
@@ -46,19 +46,57 @@ This rule aims enforce the usage of the safer operator.
 
+### `ignoreTernaryTests`
+
+Setting this option to `true` (the default) will cause the rule to ignore any ternary expressions that could be simplified by using a nullish coalesce operator.
```
```suggestion
Setting this option to `true` (the default) will cause the rule to ignore any ternary expressions that could be simplified by using a nullish coalescing operator.
```

Should the operator be named "nullish coalescing operator"?

Josh-Cena commented on packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md:
```diff
@@ -46,19 +46,61 @@ This rule aims enforce the usage of the safer operator.
+foo ?? 'a string';
+foo ?? 'a string';
+
+const foo: ?string = 'bar';
```
This is not valid TypeScript code.

JoshuaKGoldberg replied:
+1, it shouldn't be in this file. Readers could get confused about what is or isn't TS syntax.

```suggestion
```

jguddas commented:
> How about the `==` operator where `null` and `undefined` are equated (for example: `null != x ? x : 'a default value'`)?

@tduyduc great idea, can you take a look https://github.com/typescript-eslint/typescript-eslint/pull/4965/commits/1a02bdca0b9acecc9cea00d97cec102867ae8376?

tduyduc commented:
Not sure about the rule source code, but the tests look great to me! :+1:

jguddas commented:
I just added support for `MemberExpressions`, it works great, but the code is sadly running a bit out of hand 🙁

jguddas commented:
I would love some feedback on the functionality and know what besides cleaning up the code needs to be done to get this merged, are there any edge cases I have forgotten to cover?

bradzacher commented:
Hi @jguddas! This PR is in the queue of PRs to reviewed, and will be re-reviewed as soon as we are able.
https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#addressing-feedback-and-beyond

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -181,3 +277,240 @@ function isMixedLogicalExpression(node: TSESTree.LogicalExpression): boolean {
+    a.type === AST_NODE_TYPES.Identifier
+  ) {
+    return (
+      // @ts-ignore
```
Peeking at the error being suppressed here:

```
Property 'name' does not exist on type 'Identifier | BigIntLiteral | BooleanLiteral | NullLiteral | NumberLiteral | RegExpLiteral | StringLiteral'.
  Property 'name' does not exist on type 'BigIntLiteral'
```

...which is reasonable given `AST_NODE_TYPES.Literal` can be any of those. I suspect you'll want to split this if statement into one if statement for each of the two `a.type` cases.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -181,3 +277,240 @@ function isMixedLogicalExpression(node: TSESTree.LogicalExpression): boolean {
+      (!!a.value && !!b.value && a.value === b.value)
+    );
+  }
+  if (a.type === AST_NODE_TYPES.MemberExpression) {
+    return isEqualMemberExpression(a, b as typeof a);
```
It's unfortunate that TypeScript doesn't know `a.type === b.type` at this point 🤔. I'd suggest searching for a corresponding issue on github.com/Microsoft/TypeScript, creating a new one if it doesn't exist, and linking the issue you find/create here.

jguddas replied:
https://github.com/microsoft/TypeScript/issues/42384

JoshuaKGoldberg reviewed (changes requested):
You're right, this is getting to be a lot of code 😄. But that feels reasonable, this is a very tricky change you're tackling. Lots of edge cases! Nice job getting it this far 👏. 

There are also a lot of missing cases as pointed out by the TypeScript errors & codecov complaints. In good TDD form, I'd suggest you fill in all those tests so we can get a feel for what the code needs to do. Then it should be more clear what can or can't be refactored.

jguddas commented:
@JoshuaKGoldberg  I cleaned this up a bit, maybe you can take a look again.

> There are also a lot of missing cases as pointed out by the TypeScript errors & codecov complaints.

Yeah, those errors are super helpful but look worse than they are, I fixed this by adding 1 new test case and deleting one unreachable/duplicated check.

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/prefer-nullish-coalescing.test.ts:
```diff
@@ -166,6 +229,132 @@ x ?? 'foo';
+      'null == x ? y : x;',
+      'x != undefined ? x : y;',
+      'x != null ? x : y;',
+      'x == undefined  ? y : x;',
```
Nit, 😛 

```suggestion
      'x == undefined ? y : x;',
```

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/prefer-nullish-coalescing.test.ts:
```diff
@@ -71,6 +71,69 @@ x ?? 'foo';
+x === null ? x : y;
+      `,
+    ].map(code => ({
+      code: code.trim(),
```
Elsewhere, no need to `.trim()` or `as const` if TypeScript doesn't complain and the test would still pass without it.

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/prefer-nullish-coalescing.test.ts:
```diff
@@ -71,6 +71,69 @@ x ?? 'foo';
+      'null == x ? x : y;',
+      'undefined == x ? x : y;',
+      'null != x ? y : x;',
+      'undefined != x ? y : x;',
```
A few missing kinds of structures that aren't tested, don't currently, produce reports, and I think should...

* `x != undefined && x != null ? x : y`
* `x != undefined && x !== null ? x : y`
* `x !== undefined && x != null ? x : y`
* `this != undefined ? this : y;`

JoshuaKGoldberg reviewed (changes requested):
This is shaping up very nicely! 👏 

There are some more logical edge cases I can think of and commented. Also a missing tested thing.

I reviewed a bit more thoroughly this time, and can do a full depth review once those edge cases are in. I suspect you'll have to do a little bit of refactoring 🙂. Have fun!

Josh-Cena commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -330,11 +330,7 @@ function isFixableExplicitTernary({
-
-  return true;
+  return !hasNullCheck;
 }
```
Doh, isn't the logic going the other way?

JoshuaKGoldberg replied:
Err, yes, my bad 😄

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -181,3 +277,172 @@ function isMixedLogicalExpression(node: TSESTree.LogicalExpression): boolean {
+    return false;
+  }
+
+  return true;
```
oops

Josh-Cena commented:
Just... don't force push

jguddas commented:
> Just... don't force push

Josh-Cena commented:
At least that's what's said in the CONTRIBUTING guides... I'm not a maintainer/reviewer anyway.

jguddas commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -181,3 +306,54 @@ function isMixedLogicalExpression(node: TSESTree.LogicalExpression): boolean {
+  } else if (node.type === AST_NODE_TYPES.LogicalExpression) {
+    return [...getNodes(node.left), ...getNodes(node.right)];
+  }
+  return [node];
```
This line can't currently be reached, but I think it's still good to have here.

JoshuaKGoldberg replied:
> can't currently be reached

> good to have

Those two are incompatible 😛. 

Also, I'm a bit confused by this (rather vaguely named) function. Why retrieve all nodes here? Can you help me out please?

jguddas replied:
If we have `a !== null && b !== undefined` it will give us `[a, null, b, undefined]`.
This is a pretty generic function, the reason it can't be reached is due to us only using it for one specific use case.

We only use it with either a binary or logical expression due to us using `getOperator` before, which fails if we use anything else.

jguddas replied:
I will just inline `getOperator` and `getNodes`.

jguddas commented on packages/eslint-plugin/src/util/isNodeEqual.ts:
```diff
@@ -0,0 +1,77 @@
+ *   YieldExpression
+ */
+
+export function isNodeEqual(a: TSESTree.Node, b: TSESTree.Node): boolean {
```
I will think about this one.

jguddas replied:
I have thought about this, I don't like `doNodesContainSameValue` since we don't just compare the contents or value.

I however also have no idea for a better name 🤷‍♂️

jguddas commented:
This is in a presentable state now 🎉

JoshuaKGoldberg commented:
Re the force pushing: yup, it's not a blocker for review, but does make it a little harder on our end. Thanks 😄  https://github.com/typescript-eslint/typescript-eslint/issues/5170

jguddas commented on packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts:
```diff
@@ -52,15 +63,187 @@ export default util.createRule<Options, MessageIds>({
+            ) {
+              operator = '!=';
+            }
+          }
```
The other way to solve this is to put a bunch of `else { return; }` here instead of the one `if (!operator) { return; }` at the end

JoshuaKGoldberg replied:
Yeah I tried refactoring a bit and couldn't find anything significantly cleaner. 🤷

jguddas commented:
@JoshuaKGoldberg what do you think about changing the default `ignoreTernaryTests: true` to `false` in a future major release?

But anyway, let's get this merged first :smile:

JoshuaKGoldberg commented on packages/eslint-plugin/src/util/isNodeEqual.ts:
```diff
@@ -0,0 +1,47 @@
+import { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/utils';
+/*
+ * Supported:
+ *   Identifier Literal MemberExpression ThisExpression
+ *   TSNamedTupleMember TSNamespaceExportDeclaration TSNeverKeyword
+ *   TSNonNullExpression TSNullKeyword TSNumberKeyword TSObjectKeyword
+ *   TSOptionalType TSParameterProperty TSPrivateKeyword TSPropertySignature
+ *   TSProtectedKeyword TSPublicKeyword TSQualifiedName TSReadonlyKeyword
+ *   TSRestType TSStaticKeyword TSStringKeyword TSSymbolKeyword
+ *   TSTemplateLiteralType TSThisType TSTupleType TSTypeAliasDeclaration
+ *   TSTypeAnnotation TSTypeAssertion TSTypeLiteral TSTypeOperator
+ *   TSTypeParameter TSTypeParameterDeclaration TSTypeParameterInstantiation
+ *   TSTypePredicate TSTypeQuery TSTypeReference TSUndefinedKeyword TSUnionType
+ *   TSUnknownKeyword TSVoidKeyword UnaryExpression UpdateExpression
+ *   VariableDeclaration VariableDeclarator WhileStatement WithStatement
+ *   YieldExpression
+ */
```
Coming back to this (https://github.com/typescript-eslint/typescript-eslint/pull/4965#discussion_r894990290), I don't think we need this _Supported_ list either to be honest. There's not much use in having a list of several dozen node types. Folks aren't likely to read through it, and there's nothing enforcing it does work beyond the unit tests. Which are documentation enough themselves.

```suggestion
```

JoshuaKGoldberg reviewed (dismissed):
Yup, this looks great - thanks for all your hard work on this @jguddas!
